### PR TITLE
EYQB-1224 - Split out the help form into own pages

### PR DESCRIPTION
**Pull request by mrsaxby**
# Description

Please include a summary of the changes.

## Ticket number (if applicable)
https://dfedigital.atlassian.net.mcas.ms/browse/EYQB-1224

# How Has This Been Tested?

Unit tests
E2e tests 
Snap shots

# Screenshots

Please include screenshots if applicable.

# Checklist:

- [X] My code follows the standards used within this project
- [X] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests (Unit, E2E) that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/Dfe.EarlyYearsQualification.UnitTests/Helpers/StringDateHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using Dfe.EarlyYearsQualification.Web.Helpers;
+
+namespace Dfe.EarlyYearsQualification.UnitTests.Helpers;
+
+[TestClass]
+public class StringDateHelperTests
```
Comment (sam-c-dfe) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Helpers/StringDateHelperTests.cs:6]: Might be worth adding an extra test here to check it returns null if the date split length is > 2?

File: src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpProvideDetailsPageMapper.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+using Dfe.EarlyYearsQualification.Web.Mappers.Interfaces.Help;
+using Dfe.EarlyYearsQualification.Web.Models.Content.HelpViewModels;
+
+namespace Dfe.EarlyYearsQualification.Web.Mappers.Help;
+
+public class HelpProvideDetailsPageMapper() : IHelpProvideDetailsPageMapper
+{
+    public ProvideDetailsPageViewModel MapProvideDetailsPageContentToViewModel(HelpProvideDetailsPage content, string reasonForEnquiring)
+    {
+        var backButton = reasonForEnquiring == "Question about a qualification"
```
Comment (sam-c-dfe) [inline src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpProvideDetailsPageMapper.cs:11]: If this is the same string used in the controller method, might be worthing creating a constant for it?

File: src/Dfe.EarlyYearsQualification.Web/Models/Content/HelpViewModels/ProvideDetailsPageViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    
+    public string PostHeadingContent { get; init; } = string.Empty;
+    
+    public string CtaButtonText { get; init; } = string.Empty;
+
+    public string AdditionalInformationWarningText { get; init; } = string.Empty;
+
+    // text area input
+    [Required]
+    [IncludeInTelemetry]
```
Comment (sam-c-dfe) [inline src/Dfe.EarlyYearsQualification.Web/Models/Content/HelpViewModels/ProvideDetailsPageViewModel.cs:21]: I'd be tempted to remove this from here just incase they paste in email addresses or anything like that as it will end up in our logs potentially

File: tests/Dfe.EarlyYearsQualification.UnitTests/Mappers/Help/HelpGetHelpPageMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                { Label = "I have a question about a qualification", Value = "QuestionAboutAQualification" },
+                new EnquiryOption
+                { Label = "I am experiencing an issue with the service", Value = "IssueWithTheService" }
+            ]
+        };
+
+        var enquiryReasons = new List<Option>
+                             {
+                                 new() { Label = "I have a question about a qualification", Value = "QuestionAboutAQualification" },
+                                 new() { Label = "I am experiencing an issue with the service", Value = "IssueWithTheService" },
```
Comment (sam-c-dfe) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Mappers/Help/HelpGetHelpPageMapperTests.cs:50]: Could reuse the new constant in this file too?

File: tests/Dfe.EarlyYearsQualification.UnitTests/Mocks/MockContentfulServiceTests.cs [hunk 5]
```diff
@@ -663,45 +664,168 @@ public async Task GetCheckYourAnswersPage_ReturnsExpectedDetails()
...
     {
         var contentfulService = new MockContentfulService();
 
-        var result = await contentfulService.GetHelpPage();
+        var result = await contentfulService.GetGetHelpPage();
+
+        var enquiryReasons = new List<Option>
+                             {
+                                 new() { Label = "I have a question about a qualification", Value = "QuestionAboutAQualification" },
+                                 new() { Label = "I am experiencing an issue with the service", Value = "IssueWithTheService" },
```
Comment (sam-c-dfe) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Mocks/MockContentfulServiceTests.cs:678]: Constant could be used here too

File: tests/Dfe.EarlyYearsQualification.UnitTests/Services/ContentfulContentServiceTests.cs [hunk 6]
```diff
@@ -1054,41 +1055,143 @@ public async Task GetCheckYourAnswersPage_ContentfulHasNoData_ReturnsNull()
...
+    {
+        ClientMock.Setup(c =>
+                             c.GetEntriesByType(It.IsAny<string>(),
+                                                It.IsAny<QueryBuilder<HelpEmailAddressPage>>(),
+                                                It.IsAny<CancellationToken>()))
+                  .ReturnsAsync(new ContentfulCollection<HelpEmailAddressPage> { Items = [] });
+
+        var service = new ContentfulContentService(Logger.Object, ClientMock.Object, new Mock<IDateValidator>().Object);
+
+        var result = await service.GetHelpProvideDetailsPage();
```
Comment (sam-c-dfe) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Services/ContentfulContentServiceTests.cs:1190]: Should this be `var result = await service.GetHelpEmailAddressPage()`?

**Review by sam-c-dfe — CHANGES_REQUESTED** [review]

File: src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            errorSummaryLink.ElementLinkId = errorSummaryLink.ElementLinkId switch
+            {
+                nameof(FieldId.Month) => dateQuestion.MonthId,
+                nameof(FieldId.Year) => dateQuestion.YearId,
+                _ => errorSummaryLink.ElementLinkId
+            };
+        }
+
+        if (dateQuestion is not null &&
```
Comment (RobertGHippo) [inline src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs:122]: `dateQuestion` can't be null here, the parameter definition doesn't let it

File: src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            (dateQuestion.MonthError || dateQuestion.YearError) &&
+            dateQuestion.ErrorSummaryLinks is not null)
```
Comment (RobertGHippo) [inline src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs:124]: `ErrorSummaryLinks` is a non-nullable `List`: is whether it is null the correct check?

File: src/Dfe.EarlyYearsQualification.Web/Services/UserJourneyCookieService/UserJourneyCookieService.cs [hunk 9]
```diff
@@ -460,6 +463,28 @@ public string GetHasUserGotEverythingTheyNeededToday()
         }
     }
 
+    public HelpFormEnquiry GetHelpFormEnquiry()
+    {
+        lock (_lockObject)
+        {
+            EnsureModelLoaded();
+
+            return _model!.HelpFormEnquiry;
```
Comment (RobertGHippo) [inline src/Dfe.EarlyYearsQualification.Web/Services/UserJourneyCookieService/UserJourneyCookieService.cs:472]: `_model!.HelpFormEnquiry` could be null, so should the return type be `HelpFormEnqury?` ?

File: src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+using Dfe.EarlyYearsQualification.Web.Models.Content.HelpViewModels;
+using Dfe.EarlyYearsQualification.Web.Models.Content.QuestionModels;
+using Dfe.EarlyYearsQualification.Web.Models.Content.QuestionModels.Validators;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Dfe.EarlyYearsQualification.Web.Mappers.Help;
+
+public class HelpQualificationDetailsPageMapper(IPlaceholderUpdater placeholderUpdater) : IHelpQualificationDetailsPageMapper
+{
+    public QualificationDetailsPageViewModel MapQualificationDetailsContentToViewModel(QualificationDetailsPageViewModel viewModel, HelpQualificationDetailsPage content, DatesValidationResult? validationResult, ModelStateDictionary modelState)
```
Comment (RobertGHippo) [inline src/Dfe.EarlyYearsQualification.Web/Mappers/Help/HelpQualificationDetailsPageMapper.cs:15]: Quite a few warnings in this method:
unused variables;
`dateQuestion` passed into `MapDate()` might be null so should the called method handle nulls?

**Review by RobertGHippo — COMMENTED** [review]
I think the structure of this code is very good.

There are some warnings in `HelpControllerTests.cs` and `HelpQualificationDetailsPageMapperTests.cs` which could be addressed, not necessarily blockers -- mostly assignment of possibly null things to non-nullable reference types, or dereference of possibly null ref. In tests, our practice is mostly to tell it we know we expect it to be not null by using the `!` property modifier.

File: tests/Dfe.EarlyYearsQualification.UnitTests/Controllers/HelpControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,1171 @@
...
+
+        var model = resultType.Model as ConfirmationPageViewModel;
+        model.Should().NotBeNull();
+
+        model.SuccessMessage.Should().Be(content.SuccessMessage);
+    }
+
+    #region private methods for setup
+
+    private void SetUpQualificationDetailsContent()
```
Comment (RobertGHippo) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Controllers/HelpControllerTests.cs:1118]: I don't think this method is ever used?

File: tests/Dfe.EarlyYearsQualification.UnitTests/Controllers/HelpControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,1171 @@
...
+            {
+                ReasonForEnquiring = "Question about a qualification",
+            }
+        );
+
+        mockQualificationDetailsPageMapper.Setup(x => x.MapQualificationDetailsContentToViewModel(new(), content, null , new())).Returns(new QualificationDetailsPageViewModel());
+
+    }
+
+    private QualificationDetailsPageViewModel SetUpSubmission(string? qualificationName, string? awardingOrganisation,
```
Comment (RobertGHippo) [inline tests/Dfe.EarlyYearsQualification.UnitTests/Controllers/HelpControllerTests.cs:1137]: I don't think this method is ever used?

**Review by sam-c-dfe — DISMISSED** [review]

**Review by sam-c-dfe — APPROVED** [review]
